src/devices/sound/discrete.cpp

```diff
@@ -91,61 +91,70 @@ DEFINE_DEVICE_TYPE(DISCRETE, discrete_sound_device, "discrete", "Discrete Sound"
 
 struct output_buffer
 {
+	output_buffer() : source(nullptr), ptr(nullptr)
+	{
+	}
+
+	// required for use in vector (std::atomic<T> has deleted copy constructor)
+	output_buffer(output_buffer &&that) : node_buf(std::move(that.node_buf)), source(that.source), ptr(that.ptr.load(std::memory_order_relaxed)), node_num(that.node_num)
+	{
+	}
+
 	std::unique_ptr<double []>  node_buf;
-	const double                *source;
-	volatile double             *ptr;
+	const double *              source;
+	std::atomic<double *>       ptr;
 	int                         node_num;
 };
 
 struct input_buffer
 {
-	volatile const double       *ptr;               /* pointer into linked_outbuf.nodebuf */
-	output_buffer *             linked_outbuf;      /* what output are we connected to ? */
-	double                      buffer;             /* input[] will point here */
+	const double *              ptr;                // pointer into linked_outbuf.nodebuf
+	output_buffer *             linked_outbuf;      // what output are we connected to?
+	double                      buffer;             // input[] will point here
 };
 
 class discrete_task
 {
-	friend class discrete_device;
 public:
-	virtual ~discrete_task() { }
-
-	inline void step_nodes();
-	inline bool lock_threadid(int32_t threadid)
+	discrete_task(discrete_device &pdev) : m_device(pdev), m_threadid(-1), m_samples(0)
 	{
-		int expected = -1;
-		return m_threadid.compare_exchange_weak(expected, threadid, std::memory_order_release,std::memory_order_relaxed);
+		// FIXME: the code expects to be able to take pointers to members of elements of this vector before it's filled
+		source_list.reserve(16);
 	}
-	inline void unlock() { m_threadid = -1; }
+
+	void check(discrete_task &dest_task);
+	void prepare_for_queue(int samples);
+
+	static void *task_callback(void *param, int threadid);
 
 	//const linked_list_entry *list;
 	discrete_device::node_step_list_t        step_list;
 
-	/* list of source nodes */
-	std::vector<input_buffer> source_list;      /* discrete_source_node */
-
 	int task_group = 0;
 
+private:
+	void step_nodes();
+	bool process();
 
-	discrete_task(discrete_device &pdev) : m_device(pdev), m_threadid(-1)
+	bool lock_threadid(int32_t threadid)
 	{
-		// FIXME: the code expects to be able to take pointers to members of elements of this vector before it's filled
-		source_list.reserve(16);
+		int expected = -1;
+		return m_threadid.compare_exchange_strong(expected, threadid, std::memory_order_acquire, std::memory_order_relaxed);
 	}
 
-protected:
-	static void *task_callback(void *param, int threadid);
-	inline bool process();
+	void unlock()
+	{
+		m_threadid.store(-1, std::memory_order_release);
+	}
```

Comment:
Locking requires acquire semantics to ensure no task work can be reordered before it happens.  Unlocking requires release semantics to ensure no task work can be reordered after it happens.